File: backends/kotlin-ktor/src/main/kotlin/com/authorization/handler/StaffHandler.kt
/*
 * スタッフ HTTP ハンドラーモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.handler

import com.authorization.domain.staff.Condition
import com.authorization.support.AppException
import com.authorization.usecase.staff.DestroyDto
import com.authorization.usecase.staff.Interactor as StaffUC
import com.authorization.usecase.staff.UpdateRoleDto
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import kotlinx.coroutines.supervisorScope
import kotlinx.serialization.json.*
import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
import java.time.format.DateTimeFormatter

private val fmt = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")

/**
 * トランザクション内で [block] を実行し、成功時は [onId] のJSONを応答します。
 * [AppException] が発生した場合はそのステータスコードで応答します。
 *
 * newSuspendedTransaction は内部で async{}.await() を使うため、失敗時に await() で
 * 例外を受け取れても、通常のtry/catchでは周囲のコルーチンスコープがキャンセル状態のまま
 * 残ってしまい、後続の call.respond() が CancellationException で失敗する。
 * supervisorScope で失敗の伝播を子コルーチンに閉じ込めることでこれを回避する。
 */
private suspend fun ApplicationCall.respondOrAppError(onId: Long, block: suspend () -> Unit) {
    val result = supervisorScope {
        runCatching { newSuspendedTransaction { block() } }
    }
    result.fold(
        onSuccess = { respond(buildJsonObject { put("id", onId) }) },
        onFailure = { e ->
            if (e is AppException) {
                respond(HttpStatusCode.fromValue(e.statusCode), buildJsonObject { put("error", e.message) })
            } else {
                throw e
            }
        },
    )
}

/**
 * スタッフ API のハンドラーです。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
class StaffHandler(private val staffUC: StaffUC, private val cookieSecret: String) {

    /**
     * スタッフ一覧を取得します。
     *
     * @param call アプリケーションコール
     */
    suspend fun index(call: ApplicationCall) {
        val keyword  = call.request.queryParameters["keyword"]
        val roles    = call.request.queryParameters.getAll("roles")
            ?.flatMap { it.split(",") }
            ?.mapNotNull { it.trim().toIntOrNull() }
            ?: emptyList()
        val statuses = call.request.queryParameters.getAll("statuses")
            ?.flatMap { it.split(",") }
            ?.mapNotNull { it.trim().toIntOrNull() }
            ?: emptyList()
        val sort     = call.request.queryParameters["sort"]
        val sortType = call.request.queryParameters["sort_type"]
        val limit    = call.request.queryParameters["limit"]?.toIntOrNull()?.coerceAtLeast(1) ?: 10
        val page     = call.request.queryParameters["page"]?.toIntOrNull()?.coerceAtLeast(1) ?: 1
        val offset   = limit * (page - 1)

        val cond = Condition(keyword = keyword, roles = roles, statuses = statuses, offset = offset, limit = limit, sort = sort, sortType = sortType)
        val (staffs, count) = staffUC.findByConditionWithCount(cond)
        val pager = buildPager(count, limit, offset, staffs.size)

        val data = buildJsonArray {
            staffs.forEach { s ->
                add(buildJsonObject {
                    put("id",         s.id)
                    put("name",       s.name)
                    put("email",      s.email)
                    put("role",       s.role)
                    put("status",     s.status)
                    put("version",    s.version)
                    put("created_at", s.createdAt.format(fmt))
                    put("updated_at", s.updatedAt.format(fmt))
                })
            }
        }
        call.respond(buildJsonObject {
            put("data", data)
            put("pager", pager)
        })
    }

    /**
     * スタッフのロールを更新します。
     *
     * @param call アプリケーションコール
     */
    suspend fun updateRole(call: ApplicationCall) {
        val id = call.parameters["id"]?.toLongOrNull()
            ?: return call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "invalid_id") })
        val executorId = verifyStaffId(call.request.cookies["staff_id"], cookieSecret)
        val body = call.receive<JsonObject>()
        val role = body["role"]?.jsonPrimitive?.intOrNull
            ?: return call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "role_required") })
        call.respondOrAppError(id) { staffUC.updateRole(UpdateRoleDto(id = id, role = role, executorId = executorId)) }
    }

    /**
     * 論理削除されたスタッフを復元します。
     *
     * @param call アプリケーションコール
     */
    suspend fun restore(call: ApplicationCall) {
        val id = call.parameters["id"]?.toLongOrNull()
            ?: return call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "invalid_id") })
        call.respondOrAppError(id) { staffUC.restore(id) }
    }

    /**
     * スタッフを論理削除します。
     *
     * @param call アプリケーションコール
     */
    suspend fun destroy(call: ApplicationCall) {
        val id = call.parameters["id"]?.toLongOrNull()
            ?: return call.respond(HttpStatusCode.BadRequest, buildJsonObject { put("error", "invalid_id") })
        val executorId = verifyStaffId(call.request.cookies["staff_id"], cookieSecret)
        call.respondOrAppError(id) { staffUC.destroy(DestroyDto(id = id, executorId = executorId)) }
    }
}

File: backends/csharp-aspnet/src/Authorization.Api/Handler/StaffHandler.cs
/*
 * スタッフハンドラーモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
using Authorization.Api.Domain.Staff;
using Authorization.Api.Support;
using Authorization.Api.UseCase.Staff;
using static Authorization.Api.Handler.HttpHelpers;

namespace Authorization.Api.Handler;

/// <summary>スタッフハンドラーです。</summary>
public sealed class StaffHandler(StaffInteractor staffUC)
{
    /// <summary>スタッフ一覧を返します。</summary>
    /// <param name="req">HTTPリクエスト（keyword/roles/limit/page/sort/sort_typeを使用）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>一覧データとページャー情報のJSON</returns>
    public async Task<IResult> IndexAsync(HttpRequest req, CancellationToken ct)
    {
        var roles = req.Query["roles"]
            .SelectMany(v => (v ?? "").Split(','))
            .Select(s => int.TryParse(s.Trim(), out var r) ? r : (int?)null)
            .OfType<int>()
            .ToList();
        var limit = Math.Max(1, QueryInt(req, "limit") ?? 10);
        var page  = Math.Max(1, QueryInt(req, "page") ?? 1);
        if (SafeOffset(limit, page) is not int offset) return Error(400, "page_out_of_range");

        var cond = new StaffCondition
        {
            Keyword  = Query(req, "keyword"),
            Roles    = roles,
            Offset   = offset,
            Limit    = limit,
            Sort     = Query(req, "sort"),
            SortType = Query(req, "sort_type"),
        };

        var (staffs, count) = await staffUC.FindByConditionWithCountAsync(cond, ct);
        var data = staffs.Select(s => new Dictionary<string, object?>
        {
            ["id"]         = s.Id,
            ["name"]       = s.Name,
            ["email"]      = s.Email,
            ["role"]       = s.Role,
            ["status"]     = s.Status,
            ["created_at"] = DateFormat.ToMinute(s.CreatedAt),
            ["updated_at"] = DateFormat.ToMinute(s.UpdatedAt),
            ["version"]    = s.Version,
        }).ToList();

        return Results.Json(new Dictionary<string, object?>
        {
            ["data"]  = data,
            ["pager"] = Pager.Build(count, limit, offset, staffs.Count).ToJson(),
        });
    }

    /// <summary>スタッフの権限を更新します。</summary>
    /// <param name="id">スタッフID（文字列）</param>
    /// <param name="req">HTTPリクエストボディ（role）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>更新したスタッフIDのJSON。IDが不正な場合は400、roleが未指定の場合は400</returns>
    /// <exception cref="AppException">ロール値が不正な場合（400）、存在しない場合（404）</exception>
    public async Task<IResult> UpdateRoleAsync(string id, HttpRequest req, CancellationToken ct)
    {
        if (!long.TryParse(id, out var staffId)) return InvalidId();
        var body = await ReadJsonObjectAsync(req, ct);
        var role = body.Int("role");
        if (role is null) return Error(400, "role_required");

        await staffUC.UpdateRoleAsync(new StaffUpdateRoleDto(staffId, role.Value, StaffId(req)), ct);
        return Results.Json(new Dictionary<string, object?> { ["id"] = staffId });
    }

    /// <summary>スタッフの論理削除を取り消します。</summary>
    /// <param name="id">スタッフID（文字列）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>復元したスタッフIDのJSON。IDが不正な場合は400</returns>
    /// <exception cref="AppException">存在しない場合（404）</exception>
    public async Task<IResult> RestoreAsync(string id, CancellationToken ct)
    {
        if (!long.TryParse(id, out var staffId)) return InvalidId();
        await staffUC.RestoreAsync(staffId, ct);
        return Results.Json(new Dictionary<string, object?> { ["id"] = staffId });
    }

    /// <summary>スタッフを論理削除します。</summary>
    /// <param name="id">スタッフID（文字列）</param>
    /// <param name="req">HTTPリクエスト（操作者IDをクッキーから取得）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>削除したスタッフIDのJSON。IDが不正な場合は400</returns>
    /// <exception cref="AppException">存在しない場合（404）、バージョン不一致（409）</exception>
    public async Task<IResult> DestroyAsync(string id, HttpRequest req, CancellationToken ct)
    {
        if (!long.TryParse(id, out var staffId)) return InvalidId();
        await staffUC.DestroyAsync(new StaffDestroyDto(staffId, StaffId(req)), ct);
        return Results.Json(new Dictionary<string, object?> { ["id"] = staffId });
    }
}
